args/audio/systems/audiosystem.inl

```diff
@@ -0,0 +1,273 @@
+			log::error("OpenAl context failed to create");
+			openal_error();
+#if defined(AUDIO_EXIT_ON_FAIL)
+			raiseEvent<events::exit>();
+#endif
+			return;
+		}
+		// Succesfully created alcContext
+		log::info("Succesfully created openAl context");
+
+		alListener3f(AL_POSITION, 0, 0, 0);
+		alListener3f(AL_VELOCITY, 0, 0, 0);
+		ALfloat ori[] = { 0, 0, 1.0f, 0, 1.0f, 0 };
+		alListenerfv(AL_ORIENTATION, ori);
+		m_listenerPosition = position(0,0,0);
+		log::info("initialized listener!");
+
+		alDopplerFactor(1.0f);
+		alDistanceModel(AL_EXPONENT_DISTANCE);
+
+		queryInformation();
+
+		//m_sourcePositions = std::unordered_map<ecs::component_handle<audio_source>, position>();
+
+		//ARGS function binding
+
+		sourceQuery = createQuery<audio_source>();
+
+		createProcess<&AudioSystem::update>("Audio");
+		bindToEvent<events::component_creation<audio_source>, &AudioSystem::onAudioSourceComponentCreate>();
+		bindToEvent<events::component_destruction<audio_source>, &AudioSystem::onAudioSourceComponentDestroy>();
+		bindToEvent<events::component_creation<audio_listener>, &AudioSystem::onAudioListenerComponentCreate>();
+		bindToEvent<events::component_destruction<audio_listener>, &AudioSystem::onAudioListenerComponentDestroy>();
+
+		// Release context on this thread
+		alcMakeContextCurrent(nullptr);
+		mp3_audio_loader::context = std::shared_ptr<ALCcontext>(alContext);
+	}
+
+	inline AudioSystem::~AudioSystem()
+	{
+	}
+
+	inline void AudioSystem::queryInformation()
+	{
+		const ALchar* vendor = alGetString(AL_VENDOR);
+		const ALchar* version = alGetString(AL_VERSION);
+		const ALchar* renderer = alGetString(AL_RENDERER);
+		const ALchar* openALExtensions = alGetString(AL_EXTENSIONS);
+		const ALchar* ALCExtensions = alcGetString(alDevice, ALC_EXTENSIONS);
+		ALCint srate;
+		alcGetIntegerv(alDevice, ALC_FREQUENCY, 1, &srate);
+		log::info("OpenAL info:\n\n\t\t\tOpenAL information\n\tVendor: {}\n\tVersion: {}\n\tRenderer: {}\n\tOpenAl Extensions: {}\n\tALC Extensions: {}\n\tDevice samplerate: {}\n",
+			vendor, version, renderer, openALExtensions, ALCExtensions, srate);
+	}
+
+	inline void AudioSystem::update(time::span deltatime)
+	{
+		async::readonly_guard guard(m_contextLock);
+		alcMakeContextCurrent(alContext);
+
+		for (auto entity : sourceQuery)
+		{
+			auto sourceHandle = entity.get_component_handle<audio_source>();
+
+			audio_source source = sourceHandle.read();
+			position p = entity.read_component<position>();
+			position& previousP = m_sourcePositions.at(sourceHandle);
+			math::vec3 vel = previousP - p;
+			previousP = p;
+			alSource3f(source.m_sourceId, AL_POSITION, p.x, p.y, p.z);
+			alSource3f(source.m_sourceId, AL_VELOCITY, vel.x, vel.y, vel.z);
+			if (source.m_changes & 1)
+			{
+				// Pitch has changed
+				alSourcef(source.m_sourceId, AL_PITCH, source.getPitch());
+			}
+			if (source.m_changes & 2)
+			{
+				// Gain has changed
+				alSourcef(source.m_sourceId, AL_GAIN, source.getGain());
+			}
```

Comment:
same thing as before maybe think of using an enum to make this more readable and maintainable